### PR TITLE
Support TCP type runtime endpoint for kubelet

### PR DESCRIPTION
**Pull request by karataliu**
**What this PR does / why we need it**:
Currently the grpc server for kubelet and dockershim has a hardcoded endpoint: unix socket '/var/run/dockershim.sock', which is not applicable on non-unix OS.

This PR is to support TCP endpoint type besides unix socket.

**Which issue this PR fixes** 
This is a first attempt to address issue https://github.com/kubernetes/kubernetes/issues/45927

**Special notes for your reviewer**:
Before this change, running on Windows node results in:
```
Container Manager is unsupported in this build
```

After adding the cm stub, error becomes:
```
listen unix /var/run/dockershim.sock: socket: An address incompatible with the requested protocol was used.
```

This PR is to fix those two issues.

After this change, still meets 'seccomp' related issue when running on Windows node, needs more updates later.

**Release note**:

### CODE REVIEW

**Comment by yujuhong** [comment]
/cc @Random-Liu @feiskyer 

@karataliu, thanks for the PR! I have a few questions.

* What's the status of windows support w/ this PR? We don't have any testing and it's hard for me to gauge how complete this PR is. I remember seeing more changes in docker_manager.go about host config and network setting changes.
* Supporting TCP may cause security concerns since the connection could be non-local. I don't think we are ready to do that. /cc @timstclair

**Comment by yujuhong** [comment]
> Supporting TCP may cause security concerns since the connection could be non-local. I don't think we are ready to do that. /cc @timstclair

@karataliu I'd be ok with adding an `--experimental-windows` flag, and only allow TCP connections when the flag is set. I think it also aligns with the fact that windows support is in its early Alpha stage.

**Comment by karataliu** [comment]
@yujuhong Thanks for the comments.

1. I can see the concern. If the communication is limited to local only, one option is to switch to named pipe on windows, which supports more security settings.
   I've some attempts but seems grpc on windows does not work well with named pipe now. So here it is going to fallback to TCP.

2. I've pushed a new commit by extracting out two os dependent methods, so that both linux and windows build would only support specific protocol: 
   - linux will support 'unix socket' (which is the existing behaviour)
   - windows would support 'tcp' for now, and may change to named pipe or others in future
   
    Please comment if you think a new flag or other approach is preferred.

3. Since there's no testing as you mentioned, I also do not get a full map on how complete the support will be. For now I suppose the first step it's to catch up with docker manager implementation in v1.6 .
  
   We're still dealing with docker running on windows, thus code snippets in docker_manager would remain useful. This PR is to fix the issue that blocks all remaining changes.

p.s.  With this PR included, we'll get the following error during pod creating:
```
"CreatePodSandbox for pod \"s1_default(11312b63-3ebc-11e7-809f-000d3a80d419)\" failed: rpc error: code = 2 desc = failed to start sandbox container for pod \"s1\": 
Error response from daemon: {\"message\":\"security option not supported: seccomp\"}"
```

  And related code of docker_manager in v1.6:
https://github.com/kubernetes/kubernetes/blob/release-1.6.3/pkg/kubelet/dockertools/docker_manager_windows.go#L78

**Comment by feiskyer** [comment]
@k8s-bot ok to test

**Comment by yujuhong** [comment]
> I can see the concern. If the communication is limited to local only, one option is to switch to named pipe on windows, which supports more security settings.
> I've some attempts but seems grpc on windows does not work well with named pipe now. So here it is going to fallback to TCP.
> 
> I've pushed a new commit by extracting out two os dependent methods, so that both linux and windows build would only support specific protocol:
> 
> linux will support 'unix socket' (which is the existing behaviour)
> windows would support 'tcp' for now, and may change to named pipe or others in future
> Please comment if you think a new flag or other approach is preferred.

Yes, named pipe would be a better choice if grpc can support it. I am okay with Windows using TCP for now since it's an experimental/Alpha feature. The idea sounds good to me. Will review the PR later.

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -302,8 +302,8 @@ func (c *kubeletConfiguration) addFlags(fs *pflag.FlagSet) {
 	// CRI flags.
 	fs.BoolVar(&c.ExperimentalDockershim, "experimental-dockershim", c.ExperimentalDockershim, "Enable dockershim only mode. In this mode, kubelet will only start dockershim without any other functionalities. This flag only serves test purpose, please do not use it unless you are conscious of what you are doing. [default=false]")
 	fs.MarkHidden("experimental-dockershim")
-	fs.StringVar(&c.RemoteRuntimeEndpoint, "container-runtime-endpoint", c.RemoteRuntimeEndpoint, "[Experimental] The unix socket endpoint of remote runtime service.")
-	fs.StringVar(&c.RemoteImageEndpoint, "image-service-endpoint", c.RemoteImageEndpoint, "[Experimental] The unix socket endpoint of remote image service. If not specified, it will be the same with container-runtime-endpoint by default.")
+	fs.StringVar(&c.RemoteRuntimeEndpoint, "container-runtime-endpoint", c.RemoteRuntimeEndpoint, "[Experimental] The endpoint of remote runtime service. Currently unix socket is supported on Linux, and tcp is supported on windows.  Examples:'unix:///var/run/dockershim.sock', 'tcp://localhost:3735'")
```
Comment (yujuhong) [inline cmd/kubelet/app/options/options.go]: You changed the format of the endpoint to include the protocol. I don't want to break existing users. Could you make sure that endpoints without protocols can still be accepted by assuming the default protocol for each OS?

File: pkg/kubelet/dockershim/cm/container_manager_windows.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
+// +build windows
+
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (yujuhong) [inline pkg/kubelet/dockershim/cm/container_manager_windows.go]: s/2016/2017

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -577,6 +570,9 @@ func NewMainKubelet(kubeCfg *componentconfig.KubeletConfiguration, kubeDeps *Kub
 			return nil, fmt.Errorf("unsupported CRI runtime: %q", kubeCfg.ContainerRuntime)
 		}
 		runtimeService, imageService, err := getRuntimeAndImageServices(kubeCfg)
+		if err != nil {
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Good catch! :-)

**Comment by karataliu** [comment]
Thanks for the comments.

I thinks it's good to keep compatible with existing "_container-runtime-endpoint_" and "_image-service-endpoint_" flag formats, though they're still marked as '[Experimental]'.

But for windows, since the CRI support is new, suppose we could enforce the full url format as there being no 'legacy' behaviour. Otherwise it is likely to become chaos if we're to change **default** protocol type on windows in future.
For example, when switching from TCP to named PIPE, we might need to keep support for both 'localhost:6666' (exisiting behaviour), and also '\\.\pipe\PipeName' (new default).

I've pushed a new commit: add fallback support for using '/var/run/a.sock' as unix socket on linux, and also include a hint message for users to update endpoint format.  Please comment.

**Comment by yujuhong** [comment]
The new commit looks good.

Could you add some basic unit tests for the parsing functions? I think the PR will be good to go after that.

**Comment by karataliu** [comment]
unit test commit pushed.

**Review by feiskyer — APPROVED** [review]
LGTM

**Comment by feiskyer** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by yujuhong** [comment]
LGTM. Will apply the label after you squash the commits. Thanks!

**Comment by yujuhong** [comment]
> LGTM. Will apply the label after you squash the commits. Thanks!

@karataliu, to speed up the review turnaround time, feel free to send another PR for the rest of changes (e.g., fixing the security options) whenever you're ready. Thanks

**Comment by karataliu** [comment]
@yujuhong  Squashed commit pushed.  Looks fine, I'll try that. Thanks.

**Comment by feiskyer** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by yujuhong** [comment]
@karataliu thanks for working on this!

/lgtm

**Comment by yujuhong** [comment]
+ @mtaufen for approval.

**Comment by mtaufen** [comment]
componentconfig changes look fine to me
/approve

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (wking) [reply cmd/kubelet/app/options/options.go]: It's also a bit strange that the new schemes are not [registered with the IANA][1].  They seem to be [Go-isms][2].  Is it worth mentioning (here or in some other docs?) that `unix` is a `SOCK_STREAM` (vs. `unixgram`'s `SOCK_DGRAM` and `unixpacket`'s `SOCK_SEQPACKET`)?

[1]: https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml
[2]: https://golang.org/pkg/net/#Dial

**Pull request merged**